service/matching/matching_engine.go

```diff
@@ -837,6 +837,116 @@ func (e *matchingEngineImpl) listTaskQueuePartitions(request *matchingservice.Li
 	return partitionHostInfo, nil
 }
 
+func (e *matchingEngineImpl) UpdateWorkerVersioningRules(
+	ctx context.Context,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest,
+) (*workflowservice.UpdateWorkerVersioningRulesResponse, error) {
+	ns, err := e.namespaceRegistry.GetNamespace(namespace.Name(req.GetNamespace()))
+	if err != nil {
+		return nil, err
+	}
+
+	taskQueue, err := newTaskQueueID(ns.ID(), req.GetTaskQueue(), enumspb.TASK_QUEUE_TYPE_WORKFLOW)
+	if err != nil {
+		return nil, err
+	}
+
+	tqMgr, err := e.getTaskQueuePartitionManager(ctx, taskQueue, normalStickyInfo, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// we don't set updateOptions.TaskQueueLimitPerBuildId, because the Versioning Rule limits will be checked separately
+	// we don't set updateOptions.KnownVersion, because we handle external API call ordering with conflictToken
+	updateOptions := UserDataUpdateOptions{}
+	cT := req.GetConflictToken()
+
+	err = tqMgr.UpdateUserData(ctx, updateOptions, func(data *persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, bool, error) {
+		clk := data.GetClock()
+		if clk == nil {
+clk = hlc.Zero(e.clusterMeta.GetClusterID())
+		} else {
+			prevCT, err := clk.Marshal()
+			if err != nil {
+				return nil, false, err
+			}
+			if !bytes.Equal(cT, prevCT) {
+				return nil, false, serviceerror.NewFailedPrecondition(
+					fmt.Sprintf("provided conflict token %v does not match existing one %v", cT, prevCT),
+				)
+			}
+		}
+		updatedClock := hlc.Next(clk, e.timeSource)
+		hadUnfiltered := hasUnfiltered(data.GetVersioningData().GetAssignmentRules())
+		var versioningData *persistencespb.VersioningData
+		switch req.GetOperation().(type) {
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertAssignmentRule:
+			versioningData, err = InsertAssignmentRule(
+				updatedClock,
```

Comment:
all this code till the end of the switch could be extracted into getVersionData function

Reply:
i'm not sure what you mean by this. It's very specifically meant to be a part of the `UpdateUserData` `updateFn`, so that it is done in one db transaction.